### PR TITLE
Unify LoopScope and SwitchScope

### LINKED ISSUE #407: Allow case statement placement inside switch's nested blocks
#396 brings us the following test:
```c
int main()
{
    int x = 0;
    switch(x) while (0) { default: break; }
}
```

This should start working.

See `TODO[#407]` when fixing this.

### PR DESCRIPTION
**Pull request by zawodskoj**
Another step to make scopes less worse

The new BlockScope now inherits break/continue labels instead of replacing them. This allows to use BlockScope in CompoundStatement, which is good

P.S. - new tests are added to verify new codegen, `continue`-ing in switch is now possible (when there's an outer loop)

- Closes #407

### CODE REVIEW

**Comment by zawodskoj** [comment]
I'm puzzled if there's any better way to generate labels instead of NewGuid()

File: Cesium.Ast/Statements.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public record AmbiguousBlockItem(string Item1, string Item2) : IBlockItem;
 // 6.8.4 Selection statements
 public record IfElseStatement(Expression Expression, Statement TrueBranch, Statement? FalseBranch) : Statement;
 
-public record SwitchStatement(Expression Expression, Statement Body) : Statement;
+public record SwitchStatement(Expression Expression, CompoundStatement Body) : Statement;
```
Comment (kant2002) [inline Cesium.Ast/Statements.cs:32]: I honestly sad that I have to deliever bad news for you. https://github.com/ForNeVeR/Cesium/pull/398
And it is in the standard. I was quite surprised when I implementing this.

File: Cesium.CodeGen.Tests/CodeGenContinueStatementTests.cs [hunk 2]
```diff
@@ -48,4 +48,31 @@ public Task ContinueInDoWhile() => DoTest(@"int main()
 {
     continue;
 }"));
+
+    [Fact]
+    public void ContinueInSwitch() => DoesNotCompile(@"int main()
```
Comment (kant2002) [inline Cesium.CodeGen.Tests/CodeGenContinueStatementTests.cs:53]: 👍

File: Cesium.Ast/Statements.cs (hunk 1 above)
Comment (zawodskoj) [reply Cesium.Ast/Statements.cs:32]: This is the part of #396, not exactly this PR

**Comment by zawodskoj** [comment]
Rebased onto newer #396 

#399 is now implemented (except for assignment thing)

**Comment by ForNeVeR** [comment]
Please rebase on `main`.

**Comment by zawodskoj** [comment]
Rebased on main

File: Cesium.CodeGen.Tests/CodeGenContinueStatementTests.cs [hunk 3]
```diff
@@ -48,4 +48,31 @@ public Task ContinueInDoWhile() => DoTest(@"int main()
...
+{
+    switch (1)
+    {
+    default:
+        continue;
+    }
+}", "Can't break not from loop statement");
```
Comment (ForNeVeR) [inline Cesium.CodeGen.Tests/CodeGenContinueStatementTests.cs:60]: _(Minor)_ Hm, the error message here may be clearer: you actually _can_ `break` from a non-loop construct, are you? It's only impossible to `continue` outside of a loop.

Let's consider rephrasing as "Can't use continue outside of a loop construct."

File: Cesium.CodeGen/Contexts/BlockScope.cs [hunk 4]
```diff
@@ -35,8 +36,25 @@ public void DeclareFunction(string identifier, FunctionInfo functionInfo)
             : ((IDeclarationScope)Parent).GetVariable(identifier);
     }
     public IReadOnlyDictionary<string, IType> GlobalFields => ((IDeclarationScope)Parent).GlobalFields;
+
     public void AddVariable(StorageClass storageClass, string identifier, IType variable)
-        => _variables.Add(identifier, new(identifier, storageClass, variable));
+    {
+        // quirk - passing Static variables to the parent
+        // we need more tests for that
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Contexts/BlockScope.cs:43]: _(Note for myself)_ We need more issues about more tests :)

File: Cesium.IntegrationTests/switch/duff_device.c [hunk 5]
```diff
@@ -0,0 +1,18 @@
+int main()
+{
+    int j = 0;
+    int c = 42;
+
+    // assignments are broken as expressions currently
```
Comment (ForNeVeR) [inline Cesium.IntegrationTests/switch/duff_device.c:6]: _(Note for self)_ Extract to issue.

**Review by ForNeVeR — APPROVED** [review]
Great! I'll finish this.

**Pull request merged**